clientapi/routing/sendevent.go

```diff
@@ -76,6 +81,12 @@ func SendEvent(
 		}
 	}
 
+	// create a mutex for the specific user in the specific room
```

Comment:
Given that `generateSendEvent` just before this must hit internal APIs to get `prev_events`, I'm wondering if it makes sense for the mutex to come just before this call. This is probably the main place that flooding our own events could create forks.

It's probably also fine just to use `defer mutex.(*sync.Mutex).Unlock()` directly after taking out the lock, since then it's guaranteed to be released regardless of which code path we take or which errors we encounter. We may end up holding the lock for *marginally* longer that way, but on a per-user basis I wouldn't worry about that particularly.